cli: Introduce new `--format` argument (and remove `--xunit-file`)

dbaty opened the pull request:
Any output (text or xUnit) is now directed to the standard output. It
will be easier to add new formats (such as CSV or JSON, which I plan).

dbaty committed:
cli: Introduce new `--format` argument and remove `--xunit-file`

Any output (text or xUnit) is now directed to the standard output. It
will be easier to add new formats (such as CSV or JSON, which I plan).

If you used the `--xunit-file`, you must now use `--format` and
redirect the standard output to a file.

dbaty committed:
ci: Bump GitHub actions

dbaty committed:
docs: Fix markup

dbaty committed:
cli: Add CSV format

`check-branches`, `check-fixmes` and `check-futures` commands now
accept a `--format csv` argument. Example:

TiphaineLAURENT commented on src/check_oldies/check_branches.py:
```diff
@@ -41,66 +50,37 @@ def get_parser():
         err_msg = "NOK: Some branches are too old."
-        print(err_msg)
     else:
-        err_msg = ""
-        print("OK: All branches are fresh.")
-    if out:
-        print(out)
+        ok_msg = "OK: All branches are fresh."
```
Plutôt que d'initialiser des variables qu'on va modifier ou non, on pourrait avoir une petite fonction `get_messages(has_old_branches: bool) -> tuple[str, str]` ?

dbaty replied:
Mouais, je ne suis pas convaincu. :) Là, on a bout de code de 5 lignes, simple, clair, direct. Je trouve ça plus parlant qu'un appel à une fonction.

TiphaineLAURENT commented on src/check_oldies/check_fixmes.py:
```diff
@@ -55,70 +50,44 @@ def get_parser():
-    orphan_futures = annotations.get_orphan_futures(config)
-    for orphan in orphan_futures:
-        out.append(warn(orphan_str(orphan)))
-        uncolorized_out.append(orphan_str(orphan))
+    orphan_futures = check_oldies.annotations.get_orphan_futures(config)
 
-    out = os.linesep.join(out)
+    ok_msg = err_msg = ""
```
Pareil que pour check-branches

TiphaineLAURENT commented on src/check_oldies/commands.py:
```diff
@@ -1,5 +1,9 @@
-import pipes
+# FIXME (dbaty, 2024-03-28): `pipes` is deprecated since Python 3.11,
+# will be removed in 3.13. We should use `subprocess` instead (but
+# we'll have to make sure that we can do that and still support 3.8
+# and 3.9). or perhaps get rid of it?
```
3.8 and 3.9 should not be supported as they are near the end of their security cycle and close to end-of-life : https://devguide.python.org/versions/

sandre35 replied:
What are potential limitations regarding 3.8 and 3.9 ?

dbaty replied:
I am not inclined to remove support of versions that have not yet reached their end of life (however old they might be).

TiphaineLAURENT commented on src/check_oldies/output.py:
```diff
@@ -0,0 +1,90 @@
+    lines = []
+    if ok_message:
+        lines.append(ok_message)
+    if error_message:
+        lines.append(warn(error_message))
```
Pas de couleur spécifique aux erreurs pour les différencier des alertes ?

dbaty replied:
Qu'est-ce qu'une erreur, qu'est-ce qu'une alerte ?

On a 2 sortes de lignes : 
- lignes informatives : "t'as un FIXME, il a 3 jours, je t'informe, tout va bien"
- lignes d'erreur : "t'as un FIXME, il a 200 jours, c'est une erreur"

Les premières sont affichées dans la couleur par défaut du terminal. Les secondes sont affichées en rouge.

TiphaineLAURENT replied:
J'ai peut-être mal compris mais je vois ok_message, error_message et must_warn alors j'ai pensé à 3 echelons

dbaty replied:
`must_warn` indique une ligne d'erreur : une annotation ou une branche trop vieille (ou un tag "future" orphelin).

TiphaineLAURENT commented on src/check_oldies/output.py:
```diff
@@ -0,0 +1,90 @@
+    if error_message:
+        lines.append(warn(error_message))
+
+    for obj in objects:
+        lines.append(warn(obj.to_text()) if obj.must_warn else obj.to_text())
```
Peut-être plutôt extend avec une comprehension pour éviter de faire de la réallocation à tout va

dbaty replied:
Si je comprends bien, tu suggères : 

```python
lines.extend(
    warn(obj.to_text()) if obj.must_warn else obj.to_text() for obj in objects
)
```
Mouais. Je trouve ça moins lisible, mais bon, ok, pourquoi pas.

("La réallocation à tout va" : je pense que ça ne doit pas nous coûter bien cher vu la taille de la liste dans la plupart des cas. ;) )

TiphaineLAURENT commented on src/check_oldies/output.py:
```diff
@@ -0,0 +1,90 @@
+    return xml.etree.ElementTree.tostring(suite, encoding="utf-8").decode("utf-8")
+
+
+def get_formatter(format_) -> typing.Callable:
```
`output_format` plutôt que `format_` ?

dbaty replied:
c'est fait

dbaty committed:
cli: Add `--only-old` argument to all commands

By default, all commands show all annotations, branches, etc.,
including those that are recent. When exporting to a CSV format, it's
useful to filter out the recent items and only display old ones.

sandre35 commented on tests/test_check_branches.py:
```diff
@@ -113,17 +114,17 @@ def test_output_old_branches(capfd):  # capfd is a pytest fixture
 
-def test_xunit_file_generation(tmp_path):  # tmp_path is a pytest fixture
-    xunit_path = tmp_path / "xunit.xml"
+def test_xunit_file_generation(capfd):  # capfd is a pytest fixture
```
use linting instead of comment :)

dbaty replied:
c'est fait

sandre35 commented on tests/test_output.py:
```diff
@@ -0,0 +1,60 @@
+from check_oldies import output
+
+
+# `xml.etree.ElementTree.tostring()` preserves attribute order only as of
+# Python 3.8. With prior versions, we cannot easily compare XML
+# output, so don't bother.
+HAS_UNSORTED_XML_WRITE = sys.version_info[:2] < (3, 8)
```
Python minimal version supported is 3.8. So we can remove all logic link to that

dbaty replied:
Oui, dans une autre pull request où je supprimerai le support de Python 3.7 (entre autres modernisations).

dbaty commented on src/check_oldies/output.py:
```diff
@@ -0,0 +1,90 @@
+    raise ValueError(f"Unknown output format: '{format_}'")
+
+
+def printer(objects: list, output_format, **options):
```
bien vu, c'est fait

dbaty committed:
csv: Rename `filename` column as `path` in CSV output of check-fixmes

It's a (relative) path, not a mere filename.

dbaty commented:
@sandre35 et @TiphaineLAURENT :  J'ai pris en compte tous les commentaires (parmi ceux que je souhaitais prendre en compte ;) ), rebasé par rapport à master... et ai push-forcé comme un sauvageon. :) J'ai aussi corrigé divers oublis, comme la mise à jour du changelog et de la doc. La pull request m'a l'air à peu près complète, désormais, prête à être mergée.

dbaty committed:
docs: Update configuration of ReadTheDocs
